ext/drreg/drreg.c

```diff
@@ -215,10 +288,63 @@ spill_reg(void *drcontext, per_thread_t *pt, reg_id_t reg, uint slot, instrlist_
 #endif
 }
 
+#ifdef X86
+static void
+load_indirect_block(void *drcontext, per_thread_t *pt, uint slot, instrlist_t *ilist,
+                    instr_t *where, reg_id_t scratch_block_reg)
+{
+    LOG(drcontext, DR_LOG_ALL, 3, "%s @%d." PFX " %s %d\n", __FUNCTION__, pt->live_idx,
+        get_where_app_pc(where), get_register_name(scratch_block_reg), slot);
+    /* Simply load the pointer of the block to the passed register*/
+    dr_insert_read_raw_tls(drcontext, ilist, where, tls_seg, slot, scratch_block_reg);
+}
+#endif
+
+#ifdef X86
+static void
+spill_reg_indirectly(void *drcontext, per_thread_t *pt, reg_id_t reg, uint slot,
+                     instrlist_t *ilist, instr_t *where)
+{
+    reg_id_t scratch_block_reg = DR_REG_NULL;
+    LOG(drcontext, DR_LOG_ALL, 3, "%s @%d." PFX " %s %d\n", __FUNCTION__, pt->live_idx,
+        get_where_app_pc(where), get_register_name(reg), slot);
+    ASSERT(reg_is_vector_simd(reg), "not applicable register");
+    ASSERT(pt->simd_slot_use[slot] == DR_REG_NULL ||
+               reg_resize_to_opsz(pt->simd_slot_use[slot], OPSZ_64) ==
+                   reg_resize_to_opsz(reg, OPSZ_64),
+           "internal tracking error");
+    drreg_status_t res = drreg_reserve_reg_internal(
+        drcontext, DRREG_GPR_SPILL_CLASS, ilist, where, NULL, false, &scratch_block_reg);
+    if (res != DRREG_SUCCESS)
+        drreg_report_error(res, "failed to reserve tmp register");
+    ASSERT(scratch_block_reg != DR_REG_NULL, "invalid register");
+    load_indirect_block(drcontext, pt, tls_simd_offs, ilist, where, scratch_block_reg);
+    pt->simd_slot_use[slot] =
+        (reg < pt->simd_slot_use[slot]) ? pt->simd_slot_use[slot] : reg;
+    if (reg_is_strictly_xmm(reg)) {
+        opnd_t mem_opnd = opnd_create_base_disp(scratch_block_reg, DR_REG_NULL, 1,
+                                                slot * SIMD_REG_SIZE, OPSZ_16);
+        opnd_t spill_reg_opnd = opnd_create_reg(reg);
+        PRE(ilist, where, INSTR_CREATE_movdqa(drcontext, mem_opnd, spill_reg_opnd));
+    } else if (reg_is_strictly_ymm(reg)) {
+        /* TODO i#3844: support YMM here. */
+        ASSERT(false, "SIMD register not supported");
```

Comment:
An assert is not the right thing here.  This should return DRREG_ERROR_FEATURE_NOT_AVAILABLE.  Unless this should be caught at an outer user-facing function and this is an internal error??  If so please make it clear, maybe by changing the string to "YMM is not supported and should be caught by the caller" or sthg.

There are many instances of this below for ZMM and in other functions.

Reply:
"internal error:" added and a comment above the assertion.
DRREG_ERROR_FEATURE_NOT_AVAILABLE is now returned for all new _ex() functions if spill_class is ymm or zmm.